docs/iommu.md

```diff
@@ -0,0 +1,87 @@
+access the guest memory, the same way a physical IOMMU would prevent a
+physical device from accessing the physical memory. The security gain is only
+visible from a guest perspective, which does not apply if the guest we're
+taking as example is corrupted. The benefits about security can be seen in
+case of nested virtualization as it can protect the first layer guest from
+malicious guests running in L2. 
+
+As suggested by the above statement, having a virtual IOMMU also allows for
+nested device passthrough. Let's say the system has a physical IOMMU, it is
+now possible to expose a virtual IOMMU to the guest, which will be responsible
+for updating the physical DMA remapping table. This will happen through the
+VFIO framework, and will eventually allow for passing a physical device through
+multiple layers of virtualization.
+
+It is important to identify the reasons behind enabling a virtual IOMMU as it
+will impact the performance of the attached devices.
+
+## Why virtio-iommu?
+
+The Cloud Hypervisor project decided to implement the brand new virtio-iommu
+device in order to provide a virtual IOMMU to its users. The reason being the
+simplicity brought by the paravirtualization solution. By having one side
+handled from the guest itself, it removes the complexity of trapping memory
+page accesses and shadowing them. This is why the project will not try to
+implement a full emulation of a physical IOMMU.
+
+## Pre-requisites
+
+### Kernel
+
+Since virtio-iommu has landed partially into the version 5.3 of the Linux
+kernel, a special branch is needed to get things working with Cloud Hypervisor.
+By partially, we are talking about x86 specifically, as it is already fully
+functional for ARM architectures.
+
+## Usage
+
+### Create virtual IOMMU
+
+In order to expose a virtual IOMMU to the host, it is required to create a
+virtio-iommu device and expose it through the ACPI IORT table. This can be
+achieved with the simple usage of the command line option `--iommu`.
+
+```bash
+./cloud-hypervisor \
+    --cpus 1 \
+    --memory size=512M \
+    --disk path=clear-kvm.img \
+    --kernel custom-bzImage \
+    --cmdline "console=ttyS0 root=/dev/vda3" \
+    --iommu
+```
+
+### Attach devices
+
+Exposing a virtual IOMMU is the first part, but it's not really useful if no
+device is attached to it. That's why the way to expose to the guest a specific
+device as sitting behind this IOMMU is to explicitly tag it from the command
+line with the option `iommu=on`.
+
+Not all devices support this extra option, and the default value will always
+be `off` since we want to avoid the performance impact for most users who don't
+need this.
+
+Refer to the command line `--help` to find out which device support to be
+attached to the virtual IOMMU.
+
+```bash
+./cloud-hypervisor \
+    --cpus 1 \
+    --memory size=512M \
+    --disk path=clear-kvm.img,iommu=on \
+    --kernel custom-bzImage \
+    --cmdline "console=ttyS0 root=/dev/vda3" \
+    --iommu
```

Comment:
Could we have a simpler user experience by implicitly enabling the virtio IOMMU if any of the devices (parvirt'ed or directly assigned) asks for going through the IOMMU.
The `--iommu` case would still make sense for when we'll want to do device hotplug and have it go through our IOMMU, although I guess we might also dynamically add it.
Bottom line is: can we make the `--iommu` cli switch optional?

Reply:
Yes you're right, I changed the code to enable virtio-iommu creation if at least one of the devices has `iommu=on` flag.